File: frontend/types/services.ts
import type { Localized } from './i18n';

/**
 * A curated, data-driven citizen-service entry for the on-device directory.
 * This is the reusable shape behind the government-services cards — add a new
 * service by adding one object, in all three languages. The required-document
 * lists are general guidance; each card reminds the reader to confirm the exact
 * requirements at the office, in keeping with the rest of the app.
 */
export interface CitizenService {
  id: string;
  icon: string;
  title: Localized;
  /** One line: who this service is for. */
  forWhom: Localized;
  /** Commonly required documents. */
  documents: Localized[];
  /** The steps to apply, in order. */
  steps: Localized[];
  /** Where to apply. */
  where: Localized;
  /** A deadline note, or null when the service has no fixed date. */
  deadline?: Localized | null;
  /** Whether a Mee Seva centre is a typical place to complete this. */
  meeSeva?: boolean;
}

/** A nearby Mee Seva centre from the backend Places proxy. */
export interface MeeSevaCentre {
  name: string;
  address: string;
  lat: number;
  lng: number;
  openNow?: boolean | null;
  rating?: number | null;
  distanceKm: number;
}

/** GET /places/mee-seva */
export interface MeeSevaResponse {
  enabled: boolean;
  error?: string;
  results: MeeSevaCentre[];
}

/** GET /places/config */
export interface MapsConfig {
  enabled: boolean;
  browserKey: string | null;
}

/**
 * Government-service types — mirror the collector API responses exactly
 * (GET /services, /services/{id}, /services/{id}/history, /search, /stats, /health).
 * No `any`; every backend field is typed.
 */

export interface GovServiceForm {
  title: string | null;
  url: string;
}

export interface GovServiceFAQ {
  question: string;
  answer: string;
}

export interface GovServiceContact {
  department?: string | null;
  phone?: string[];
  email?: string[];
  address?: string | null;
  website?: string | null;
}

/** One government service (collector `ServiceOut`). */
export interface GovService {
  id: number;
  service_name: string;
  department: string | null;
  state: string | null;
  district: string | null;
  language: string | null;
  description: string | null;
  eligibility: string[];
  required_documents: string[];
  application_steps: string[];
  fees: string | null;
  processing_time: string | null;
  official_application_url: string | null;
  official_notification_url: string | null;
  forms: GovServiceForm[];
  faq: GovServiceFAQ[];
  contact: GovServiceContact;
  version: number;
  last_updated: string | null;
  source_url: string | null;
}

/** A search hit adds a relevance score (collector `ServiceHit`). */
export interface GovServiceHit extends GovService {
  score: number;
}

export type SearchMode = 'keyword' | 'semantic' | 'hybrid';

/** GET /search */
export interface GovSearchResponse {
  query: string;
  mode: SearchMode;
  count: number;
  results: GovServiceHit[];
}

/** GET /services (paginated) */
export interface PaginatedServices {
  total: number;
  limit: number;
  offset: number;
  results: GovService[];
}

/** GET /services/{id}/history */
export interface GovServiceVersion {
  version: number;
  content_hash: string | null;
  captured_at: string | null;
  snapshot: Record<string, unknown>;
}

/** GET /stats */
export interface GovStats {
  total_services: number;
  by_state: Record<string, number>;
  by_language: Record<string, number>;
}

/** GET /health */
export interface GovHealth {
  status: string;
  search_mode?: string;
  embedding_model?: string;
}

/** POST /crawl */
export interface CrawlResponse {
  status: string;
  sources: string[] | string | null;
}

/** POST /chat (placeholder integration; backend endpoint prepared separately). */
export interface ChatRequest {
  question: string;
  lang?: string;
  state?: string;
  top_k?: number;
}

export interface ChatCitation {
  service_name: string;
  source_url?: string | null;
  official_application_url?: string | null;
}

export interface ChatResponse {
  answer: string;
  grounded: boolean;
  citations?: ChatCitation[];
  used_services?: GovServiceHit[];
}

/** Query params for listing / searching. */
export interface ServiceListParams {
  state?: string;
  department?: string;
  language?: string;
  limit?: number;
  offset?: number;
}

export interface ServiceSearchParams {
  q: string;
  mode?: SearchMode;
  limit?: number;
  state?: string;
  department?: string;
  language?: string;
}

/** Back-compat alias (previous name used by services/gov-services.service.ts). */
export type GovServiceSearchResponse = GovSearchResponse;

File: backend/app/schemas/services.py
"""
Schemas for government services sourced from the Sahayak Data Collector.

These mirror the collector's `ServiceOut` / `SearchResponse` / `PaginatedServices`
/ `StatsResponse` shapes exactly, so the main backend can proxy the collector's
responses through unchanged and the frontend gets one stable, typed contract.
"""

from typing import Any, Literal

from pydantic import BaseModel, Field


class GovServiceForm(BaseModel):
    title: str | None = None
    url: str


class GovServiceFAQ(BaseModel):
    question: str
    answer: str


class GovServiceContact(BaseModel):
    department: str | None = None
    phone: list[str] = Field(default_factory=list)
    email: list[str] = Field(default_factory=list)
    address: str | None = None
    website: str | None = None


class GovService(BaseModel):
    """A single government service / scheme (collector `ServiceOut`)."""

    id: int
    service_name: str
    department: str | None = None
    state: str | None = None
    district: str | None = None
    language: str | None = None
    description: str | None = None
    eligibility: list[str] = Field(default_factory=list)
    required_documents: list[str] = Field(default_factory=list)
    application_steps: list[str] = Field(default_factory=list)
    fees: str | None = None
    processing_time: str | None = None
    official_application_url: str | None = None
    official_notification_url: str | None = None
    forms: list[GovServiceForm] = Field(default_factory=list)
    faq: list[GovServiceFAQ] = Field(default_factory=list)
    contact: GovServiceContact = Field(default_factory=GovServiceContact)
    version: int = 1
    last_updated: str | None = None
    source_url: str | None = None


class GovServiceHit(GovService):
    """A search result adds a relevance score (collector `ServiceHit`)."""

    score: float


class PaginatedServices(BaseModel):
    total: int
    limit: int
    offset: int
    results: list[GovService]


class GovServiceSearchResponse(BaseModel):
    query: str
    mode: Literal["keyword", "semantic", "hybrid"]
    count: int
    results: list[GovServiceHit]


class GovServiceVersion(BaseModel):
    version: int
    content_hash: str | None = None
    captured_at: str | None = None
    snapshot: dict[str, Any] = Field(default_factory=dict)


class GovStats(BaseModel):
    total_services: int
    by_state: dict[str, int] = Field(default_factory=dict)
    by_language: dict[str, int] = Field(default_factory=dict)


class GovChatRequest(BaseModel):
    question: str = Field(..., min_length=2)
    lang: str = "en"
    state: str | None = None
    top_k: int = Field(4, ge=1, le=10)


class GovChatCitation(BaseModel):
    service_name: str
    source_url: str | None = None
    official_application_url: str | None = None


class GovChatResponse(BaseModel):
    answer: str
    grounded: bool
    citations: list[GovChatCitation] = Field(default_factory=list)
    used_services: list[GovServiceHit] = Field(default_factory=list)
